ingest/src/main/java/com/microsoft/azure/kusto/ingest/IngestClientImpl.java

```diff
@@ -253,8 +270,46 @@ public IngestionResult ingestFromResultSet(ResultSetSourceInfo resultSetSourceIn
         }
     }
 
+    protected void validateEndpointServiceType() throws IngestionServiceException, IngestionClientException {
+        if (StringUtils.isBlank(endpointServiceType) && resourceManager != null) {
+            endpointServiceType = resourceManager.getServiceType();
+        }
+        if (!DATA_MANAGEMENT_SERVICE_TYPE.equals(endpointServiceType)) {
+            String message = String.format(WRONG_ENDPOINT_MESSAGE, DATA_MANAGEMENT_SERVICE_TYPE, endpointServiceType);
+            if (StringUtils.isNotBlank(generateEndpointSuggestion())) {
+                message = String.format(message + ": '%s'", suggestedEndpointUri);
+            } else {
+                message += ".";
+            }
+            throw new IngestionClientException(message);
+        }
+    }
+
+    private String generateEndpointSuggestion() {
+        if (suggestedEndpointUri != null) {
+            return suggestedEndpointUri;
+        }
+        // The default is not passing a suggestion to the exception
+        suggestedEndpointUri = "";
+        if (StringUtils.isNotBlank(connectionDataSource)) {
+            URIBuilder engineUri;
+            try {
+                engineUri = new URIBuilder(connectionDataSource);
+                engineUri.setHost(INGEST_PREFIX + engineUri.getHost());
+                suggestedEndpointUri = engineUri.toString();
+            } catch (URISyntaxException e) {
+                log.error("Couldn't parse connectionDataSource '{}', so no suggestion can be made.", connectionDataSource, e);
+            }
+        }
+        return suggestedEndpointUri;
+    }
+
+    protected void setConnectionDataSource(String connectionDataSource) {
```

Comment:
here and with the WRONG_ENDPOINT_MESSAGE - I think u confused 'protected' with private-package (=no modifier)

Reply:
The difference between protected and default is whether subclasses in another package have access to that entity.

There's a fundamental philosophical difference between C# and Java. The former blocks method overriding by default, and you need to mark the method as virtual. See #3 in https://blog.overops.com/c-vs-java-the-top-5-features-java-developers-miss-in-c/

Further, the number of final classes in a Java project is usually much lower than the number of C# classes that are sealed (their equivalent). Java is generally developed with inheritance allowed by default for future extensibility, and C# takes the opposite approach of planning ahead the classes that you expect will need to be inherited from, and sealing everything else. See https://stackoverflow.com/a/37346728/5898291 for pros and cons of each philosophy.

The result of the above is that Java projects tend to use "protected" much more often than the default modifier to allow for subclasses in any package access to the entity (as discussed in https://stackoverflow.com/a/6470746/5898291).

Lastly, consider that the current package paradigm in our Java SDK is set up in a way that's far from ideal (see https://stackoverflow.com/questions/3226282/are-there-best-practices-for-java-package-organization for best practices on this). Currently, all business logic classes are under the same package. So package and default access would both work the same today in this repo. But if we start using better package organization practices, it might start blocking entities from being accessed in subclasses where it wasn't intended.

In these particular cases, I intentionally wanted this field and method to be accessible to subclasses regardless of project. Do you think they should be blocked from subclasses in a different project?

Reply:
no i dont
Thanks for the complete summary